### PR TITLE
add-web-push

### PR DESCRIPTION
**Pull request by henrikav**
First shell of web push up

### CODE REVIEW

File: app/api/register-push/route.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        console.error(error);
+        throw NextResponse.json(
+            { error: 'Internal server error '},
+            {status: 500}
+        );
+    }
+}
+
+
+export async function GET(_: NextRequest) {
```
Comment (dymbe) [inline app/api/register-push/route.ts]: Would move this to a different folder, as it does not GET a push-registration. For example `app/api/push`? Would consider making this a POST-endpoint instead, as this makes more sense for REST (you're creating a push-notification, not getting it), although you could keep this a GET if that makes testing easier

File: app/api/register-push/route.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import {deleteSubscriptionFromDb, getSubscriptionsFromDb, saveSubscriptionToDb } from '@/app/utils/in-memory-db'
```
Comment (dymbe) [inline app/api/register-push/route.ts]: Could you rename this folder to something like `push-subscriptions` or just `subscriptions` to make the API more RESTful?

File: app/api/register-push/route.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        throw NextResponse.json(
+        { error: 'Internal server error '},
+        {status: 500}
+        );
+    }
+}
+
+export async function DELETE(request: NextRequest) {
+    try {
+        const subscriptionToDelete: PushSubscription | undefined = await request.json();
```
Comment (dymbe) [inline app/api/register-push/route.ts]: Is it possible for `subscriptionToDelete` to be `undefined` here? I think `request.json()` will just throw an error if it has no data

File: app/utils/in-memory-db.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import {PushSubscription as DatabasePushSubscription} from 'web-push';
```
Comment (dymbe) [inline app/utils/in-memory-db.ts]: I would consider moving the util-folder out of the app-router folder

File: public/manifest.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (dymbe) [inline public/manifest.json:1]: This file probably needs some more work, but it's fine to not prioritize it right now. Would create an issue for it though

File: public/service.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+self.addEventListener('push', async (event) => {
```
Comment (dymbe) [inline public/service.js]: could you rename this to `service-worker.js`? Or `sw.js`. It's not obvious to me that this is a service-worker

File: app/page.tsx [hunk 7]
```diff
@@ -1,11 +1,88 @@
...
 
 const Home = () => {
-  return (
-    <main>
-        <Box>Hello world</Box>
-    </main>
-  )
+    const [hasActivePushSubscription, setHasActivePushSubscription] = useState<boolean>();
+
+    useEffect(() => {
```
Comment (dymbe) [inline app/page.tsx]: What do you think about merging the useEffect calls here?

**Review by dymbe — CHANGES_REQUESTED** [review]
Seems to work fine on Firefox and Mac, although it does not allow me to send the same notification twice (I assume that's a feature not a bug?). I have some suggestions, also I think you should run the formatter

File: public/manifest.json (hunk 5 above)
Comment (henrikav) [reply public/manifest.json:1]: Done

File: app/api/register-push/route.ts (hunk 3 above)
Comment (henrikav) [reply app/api/register-push/route.ts]: If the request body is empty or doesn't contain valid JSON data, the await request.json() call can result in undefined

File: app/api/register-push/route.ts (hunk 1 above)
Comment (henrikav) [reply app/api/register-push/route.ts]: I agree. Moved into push-notification folder but kept as GET.

File: app/api/register-push/route.ts (hunk 2 above)
Comment (henrikav) [reply app/api/register-push/route.ts]: Done

File: app/utils/in-memory-db.ts (hunk 4 above)
Comment (henrikav) [reply app/utils/in-memory-db.ts]: Done

File: public/service.js (hunk 6 above)
Comment (henrikav) [reply public/service.js]: Done

File: app/page.tsx (hunk 7 above)
Comment (henrikav) [reply app/page.tsx]: I agree, done.

File: README.md [hunk 8]
```diff
@@ -1,4 +1,5 @@
 # flood-frontend
+
 =======
```
Comment (dymbe) [inline README.md]: This looks like the leftovers of a merge-conflict

**Review by dymbe — CHANGES_REQUESTED** [review]
Looks good as a PoC for now, except for one little nitpick

**Review by dymbe — APPROVED** [review]

**Pull request merged**